### ISSUE TITLE
bobyqa for glmmTMB

### ISSUE THREAD

**Issue by jprice80** [glmmTMB/glmmTMB#359]
Although nlminb is currently being used, adding the bobyqa optimizer may provide an alternative routine to consider in the event of model convergence problems.  Is the bobyqa optimization routine something the authors would consider implementing?

**Comment by bbolker** [comment]
While providing the capability for alternative optimizers is something I think would be a good idea, and not difficult for any structural reason, it's not clear that it makes much sense to use BOBYQA (a derivative-free optimizer) since one of TMB/glmmTMB's great strengths is that it does accurate, fast computation of gradients -- so using an optimization method that ignores them would probably be sub-optimal ...

**Comment by jprice80** [comment]
I see.  Thanks.  Would something like L-BFGS-B be more applicable future routine to incorporate? The nlminb routine may be so similar to L-BFGS-B that it may not be worth the time at this point but I'm uncertain.

**Comment by bbolker** [comment]
Yes. Actually, at present glmmTMB doesn't do any constrained estimation (all of the parameters are estimated on transformed/unconstrained scales, e.g. standard deviations are fitted on the log scale), so we need not be limited to optimizers that can handle box constraints.  So BFGS, CG from optim() are available, as well as a variety from optimx ... see https://github.com/kaskr/adcomp/issues/73 for comparison ...

**Comment by skanskan** [comment]
So, what is the fastest optimizer options we can use with glmmTMB for a logistic model with random effects?

**Comment by bbolker** [comment]
At present use of `nlminb()` is hard-coded: https://github.com/glmmTMB/glmmTMB/blob/master/glmmTMB/R/glmmTMB.R#L857

I don't think it would hurt to allow more flexibility, but keep in mind:

- glmmTMB's big strength is automatic/efficient computation of gradients, so there would be little point in switching to a derivative-free optimizer like bobyqa.
- glmmTMB fits on an unconstrained scale (i.e., with all parameters on a scale where no constraints are needed), so box-constrained optimizers are not necessary.

Do you have any particular reason to believe that your models would benefit from a different optimizer?  That is, is there a reason that flexibility in optimizer should be a high priority for development efforts?

**Comment by skanskan** [comment]
I just wanted to find the fastest option.

**Comment by jamesdalg** [comment]
Just putting in my two cents that making a small wrapper for optimx convergence methods could be helpful in various circumstances. More flexibility could never hurt and we all know there are great, fast methods that glmmTMB does well, but sometimes with a tricky problem, it might be nice to try something different if the model is having issues converging and see if other optimizers might be suited for a particular problem. There's plenty to do on the package, but it could be nice.

**Comment by raffaelemancuso** [comment]
> Just putting in my two cents that making a small wrapper for optimx convergence methods could be helpful in various circumstances. More flexibility could never hurt and we all know there are great, fast methods that glmmTMB does well, but sometimes with a tricky problem, it might be nice to try something different if the model is having issues converging and see if other optimizers might be suited for a particular problem.

I came up with this wrapper for optimx.

```
myoptmix <- function(...) {
  ret <- optimx::optimr(...)
  # glmmTMB automatically handles output from optim(), by renaming the value component to objective
  mask <- names(ret) == "value"
  names(ret)[mask] <- "objective"
  return(ret)
}

control_nvm <- glmmTMB::glmmTMBControl(
  parallel = nt,
  optimizer = myoptmix,
  optArgs = list(method = "nvm")
)
```

Then you can pass `control` to the `control` parameter of glmmTMB.

I found that the `lbfgsb3c` method is quite fast, while the `nvm` is a bit slower but helpful in case the model fails to converge.
